### PR TITLE
fix: only inherit workspace package table if the new package is a member

### PR DESCRIPTION
**Pull request by hi-rustin**
### What does this PR try to resolve?

part of https://github.com/rust-lang/cargo/issues/13235

We should not inherit from the workspace if the new package is in the exclude list.

- [x] Updated the test case to cover it.
- [x] Do not inherit the workspace package table if it is not a member of the workspace.

### How should we test and review this PR?

Check out the unit tests.

### Additional information

None

### CODE REVIEW

**Review by hi-rustin — COMMENTED** [review]
🔢 Self-check

File: src/cargo/ops/cargo_new.rs [hunk 1]
```diff
@@ -955,8 +963,52 @@ fn update_manifest_with_inherited_workspace_package_keys(
...
+        workspace_document["workspace"]["members"] = toml_edit::value(array);
+    }
+
+    write_atomic(
+        &root_manifest_path,
+        workspace_document.to_string().to_string().as_bytes(),
+    )
+}
+
+fn get_display_path_and_check_exclusion(
```
Comment (epage) [inline src/cargo/ops/cargo_new.rs:1007]: `and`s in function names generally imply too much is being coupled together.  What would it look like if we made this two separate functions?

And if you are making changes, can you put the refactor into a separate commit?

File: src/cargo/ops/cargo_new.rs [hunk 2]
```diff
@@ -955,8 +963,52 @@ fn update_manifest_with_inherited_workspace_package_keys(
 fn update_manifest_with_new_member(
     root_manifest_path: &Path,
     workspace_document: &mut toml_edit::Document,
-    package_path: &Path,
+    display_path: String,
```
Comment (epage) [inline src/cargo/ops/cargo_new.rs:966]: nit: to make the caller more refactor friendly, this should probably be a `&str`

File: src/cargo/ops/cargo_new.rs [hunk 3]
```diff
@@ -816,18 +816,24 @@ fn mk(config: &Config, opts: &MkOptions<'_>) -> CargoResult<()> {
...
                 .get("workspace")
                 .and_then(|workspace| workspace.get("package"))
                 .and_then(|package| package.as_table())
             {
-                update_manifest_with_inherited_workspace_package_keys(
-                    opts,
-                    &mut manifest,
-                    workspace_package_keys,
-                )
+                if !exclude {
```
Comment (epage) [inline src/cargo/ops/cargo_new.rs:829]: Double-negatives aren't ideal.  Could we change the return value to `is_workspace_member`?

Comment (epage) [inline src/cargo/ops/cargo_new.rs:829]: Why is the `!exclude` check inside of the `if`?  Seems like from the perspective of trying to describe this code, we shouldn't try to look stuff up in the workspace if this is a workspace member.

Putting it on the outside will also make it easier to avoid bugs as we add new sections to inherit

File: src/cargo/ops/cargo_new.rs [hunk 4]
```diff
@@ -839,12 +845,14 @@ fn mk(config: &Config, opts: &MkOptions<'_>) -> CargoResult<()> {
                 manifest["lints"] = toml_edit::Item::Table(table);
```
Comment (epage) [inline src/cargo/ops/cargo_new.rs:845]: We are not excluding `workspace.lints`

**Comment by epage** [comment]
> part of #13235

If this doesn't close that issue, then what is left after this?

File: src/cargo/ops/cargo_new.rs (hunk 3 above)
Comment (hi-rustin) [reply src/cargo/ops/cargo_new.rs]: Moved out.

**Comment by hi-rustin** [comment]
GitHub is down.

File: src/cargo/ops/cargo_new.rs (hunk 1 above)
Comment (hi-rustin) [reply src/cargo/ops/cargo_new.rs]: Separated

**Comment by hi-rustin** [comment]
GitHub is down.

> If this doesn't close that issue, then what is left after this?

This is another bug in that issue. 

```console
❯ tree . --gitignore
.
├── Cargo.toml
├── crates
│   └── crate1
│       ├── Cargo.toml
│       └── src
│           └── main.rs
├── data
│   └── test2
│       ├── Cargo.toml
│       └── src
│           └── main.rs
└── src
    └── main.rs

8 directories, 6 files
```

The root workspace:

```toml
[package]
name = "test1"
version = "0.1.0"
edition = "2021"

[workspace]
resolver = "2"
members = ["crates/*", "data/test2"]
exclude = ["data/**"]


[workspace.package]
version = "0.1.0"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]

```

If you create a package inside the data directory, it will inherit the version from the workspace and add the package into the members list, even if it should be ignored by the exclude glob.

File: src/cargo/ops/cargo_new.rs [hunk 5]
```diff
@@ -993,46 +1042,9 @@ fn update_manifest_with_new_member(
                 .as_str()
                 .with_context(|| format!("invalid non-string exclude path `{}`", member))?;
             if pat == display_path {
```
[the remark this answers is not in the document]
Comment (epage) [reply src/cargo/ops/cargo_new.rs:1044]: According to #6009, we don't support globs on exclude...

**Comment by epage** [comment]
@bors r+

This fixes a definitive problem and we can work on the question of glob excludes separate from this

File: src/cargo/ops/cargo_new.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (epage) [reply src/cargo/ops/cargo_new.rs:1044]: Is the problem with globs or that `exclude` is a path prefix?

[the remark this answers is not in the document]
Comment (hi-rustin) [reply src/cargo/ops/cargo_new.rs:1044]: > According to #6009, we don't support globs on exclude...

Oh, I forgot that 🤣 



> Is the problem with globs or that `exclude` is a path prefix?

No. It works well.
